File: graph/strongly_connected_components/java/SCC.java
import java.util.*;

public class SCC {
  public static int[] scc(int[][] graph) {
    List<Integer> to = topologicalOrder(graph);
    return getCC(graph, to);
  }

  private static List<Integer> topologicalOrder(int[][] graph) {
    List<Integer> order = new ArrayList<Integer>();
    int n = graph.length;
    boolean[] visited = new boolean[n];
    for (int u = 0; u < n; u++) {
      if (!visited[u]) {
        topologicalDFS(graph, u, visited, order);
      }
    }
    return order;
  }

  private static void topologicalDFS(int[][] graph, int u, boolean[] visited, List<Integer> l) {
    visited[u] = true;
    for (int v : graph[u]) {
      if (!visited[v]) {
        topologicalDFS(graph, v, visited, l);
      }
    }
    l.add(u);
  }

  private static int[] getCC(int[][] graph, List<Integer> order) {
    int n = graph.length;
    int cc_cnt = 0;
    int[] cc = new int[n];
    boolean[] visited = new boolean[n];
    for (int u : order) {
      if (!visited[u]) {
        cc[u] = cc_cnt++;
        ccDFS(graph, u, visited, cc);
      }
    }
    return cc;
  }

  private static void ccDFS(int[][] graph, int u, boolean[] visited, int[] cc) {
    visited[u] = true;
    for (int v : graph[u]) {
      if (!visited[v]) {
        cc[v] = cc[u];
        ccDFS(graph, v, visited, cc);
      }
    }
  }
}

File: graph/strongly_connected_components/py/scc.py
def scc(g):
    to = topological_order(g)
    return cc(g, to)


def topological_order(g):
    l = []
    visited = {}
    for u in g.keys():
        visited[u] = False
    for u in g.keys():
        if not visited[u]:
            topological_dfs(g, u, visited, l)
    return l

def topological_dfs(g, u, visited, l):
    visited[u] = True
    for v in g[u]:
        if not visited[v]:
            topological_dfs(g, v, visited, l)
    l.append(u)

def cc(g, vertex_order):
    c = {}
    cnt = 0
    visited = {}
    for u in vertex_order:
        visited[u] = False
    for u in vertex_order:
        if not visited[u]:
            c[u] = cnt
            cc_dfs(g, u, visited, c)
            cnt += 1
    return c

def cc_dfs(g, u, visited, c):
    visited[u] = True
    for v in g[u]:
        if not visited[v]:
            c[v] = c[u]
            cc_dfs(g, v, visited, c)
